### PR TITLE
fixes #67, fixes #1: Migrate to Jersey 2 and Swagger

### LINKED ISSUE #1: Migrate to Swagger
...Migrated from [Ignite Jira OF-1065](https://issues.igniterealtime.org/browse/OF-1065)...

Right now, is the documentation hardly to maintain. I would like to migrate it to Swagger to have a clear and extendable documentation.

### LINKED ISSUE #67: Migrate Jersey from 1.* to 2.*
The latest release of Jersey1 (com.sun.jersey) is now over 4 years old and contains deprecated and vulnerable dependencies, like Jackson 1.9.2. It is not under active development.
Jersey2 has been moved to the org.glassfish.jersey.* namespace. It contains somewhat different artifacts but all the core functionality has been migrated. The upgrade path doesn't seem all that cumbersome.

Consider moving from Jersey1 to Jersey2.

### PR DESCRIPTION
**Pull request by guusdk**
This updates the Jersey dependency from 1.x to 2.x.

This commit requires OF-2352 / https://github.com/igniterealtime/Openfire/pull/1945 which means that this PR will _not_ work with 4.7.0 beta (but hopefully will with 4.7.0 non-beta).

### CODE REVIEW

**Comment by Redor** [comment]
wow, thats a good one. I have to take a bit of time to test it.

**Comment by Fishbowler** [comment]
Rebased.

**Comment by Fishbowler** [comment]
Testing on 4.7.0 master looks stable. Read and Write requests have identical behaviour and body content as master.

**Comment by guusdk** [comment]
I've now added conversion to Swagger. It is fully functional, but not all documentation has been ported yet (work in progress).

image

**Comment by guusdk** [comment]
I think I've added documentation for all endpoints. This is ready for review.

**Comment by guusdk** [comment]
Note that the Swagger documentation that's added in this PR can't be used 'offline'. We still need to replace the static documentation (in readme.html) with something that's generated during compilation/packaging. This effort is tracked as issue #71.

I'm not sure if I get around to doing that soon (the holidays have come to an end :) ). Is this PR acceptable as-is?

**Comment by Fishbowler** [comment]
This is great! I've even found a few new bugs using the swagger tool :)
Is there a way to set defaults for the XML Elements that are POST bodies? We could chip away at that in future PRs.
Also, any idea how to hide the /application.wadl entries in the default section at the bottom?

**Comment by Fishbowler** [comment]
All of the descriptions look accurate. I've added one tiny tweak to mirror the same constraint on usergroup updates.

**Comment by guusdk** [comment]
> Is there a way to set defaults for the XML Elements that are POST bodies?

I don't recall that we can set defaults, but I think we can have more meaningful examples (with more valid-looking data than 'string').

> Also, any idea how to hide the /application.wadl entries in the default section at the bottom?

I tried & failed. It _should_ be easy - but disabling wadl support (which should prevent that documentation to be added didn't actually seem to disable wadl support. I'd be happy for that to be done, as I suspect that the endpoints won't be functional anyway.

**Comment by GregDThomas** [comment]
FWIW, the values in the page default to the examples you configure in the code.

**Comment by guusdk** [comment]
@Redor what do you think? For the 4.7.0 release of Openfire, we will need to publish a new version of this plugin. I'd love to include this PR.

**Comment by Redor** [comment]
It does look good, but didn't had the time to test it. (I will be tomorrow back from vacation). Will review it ASAP

File: src/java/org/jivesoftware/openfire/plugin/rest/AuthFilter.java [hunk 1]
```diff
@@ -33,15 +39,13 @@
...
-     * (non-Javadoc)
-     * 
-     * @see
-     * com.sun.jersey.spi.container.ContainerRequestFilter#filter(com.sun.jersey
-     * .spi.container.ContainerRequest)
-     */
     @Override
-    public ContainerRequest filter(ContainerRequest containerRequest) throws WebApplicationException {
+    public void filter(ContainerRequestContext containerRequest) throws IOException {
+        if (containerRequest.getUriInfo().getRequestUri().getPath().equals("/plugins/restapi/v1/openapi.yaml")) {
```
Comment (Redor) [inline src/java/org/jivesoftware/openfire/plugin/rest/AuthFilter.java:44]: Should we exclude it? In swagger UI we can set the secret key. I think we should make fewer "exceptions" in the authentication part, which is crucial.

**Comment by Redor** [comment]
ServiceException seems not to work, to show specific Error and status codes. E.g. instead of a "bad request", I'm getting always 500. Probably the handling is slightly different in jersey 2

**Comment by Fishbowler** [comment]
Another problem on a similar vein.
I've used the openapi.yaml to import as anew collection into Insomnia, which defaults to the JSON representation.
Something looks amiss with the collection-type properties.

`POST http://localhost:9090/plugins/restapi/v1/users`

```
{
  "username": "joseph",
  "name": "Joseph",
  "email": "joseph@example.org",
  "password": "secret",
  "properties": [
    {
      "key": "somekey",
      "value": "somevalue"
    }
  ]
}
```

> Unrecognized field "properties" (class org.jivesoftware.openfire.plugin.rest.entity.UserEntity), not marked as ignorable (5 known properties: "name", "email", "property", "username", "password"])
> at [Source: (org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream); line: 6, column: 18] (through reference chain: org.jivesoftware.openfire.plugin.rest.entity.UserEntity["properties"])

Using XML doesn't encounter the same issue. Something to do with the JSON representation not happy with the `XmlElementWrapper`? https://github.com/guusdk/openfire-restAPI-plugin/blob/67_Migrate-to-Jersey-2/src/java/org/jivesoftware/openfire/plugin/rest/entity/UserEntity.java#L140

**Comment by guusdk** [comment]
> ServiceException seems not to work, to show specific Error and status codes. E.g. instead of a "bad request", I'm getting always 500.

The RestExceptionMapper didn't automatically register. I have now modified the code to do that manually which fixes the problem.

**Comment by guusdk** [comment]
> Another problem on a similar vein.

Do you know if that happened before this PR? If so, I'd rather take on this issue in a separate effort.

**Comment by Fishbowler** [comment]
> Do you know if that happened before this PR? 

Tested. It's new :(

**Comment by guusdk** [comment]
Am I right to conclude that it affects POSTs, but not GETs?

**Comment by guusdk** [comment]
Also, the JSON example titled "REST API Version 1.3.0 and later - Payload Example 3 (available parameters):" in the version prior to my changes was already not working, right? If not, then I'm not going to reinstate that.

**Comment by guusdk** [comment]
I think I've now fixed this for all occurrences of `XmlElementWrapper`. I've compared JSON based retrieval and posts with version 1.6.0 of the plugin, and I _think_ that with this update, we retain the existing behavior in 1.6.0.

**Comment by Fishbowler** [comment]
I'm seeing working JSON, and the lovely custom errors when I do the wrong thing. :+1: :+1: 
Will spend a little while longer testing this time!

**Comment by Fishbowler** [comment]
I've tested everything except the MUC permissions endpoints and the legacy UserService, and pushed one additional commit to add another annotation.
Everything JSON is working - I've not tested much in XML, as it's likely safer.

**Comment by Redor** [comment]
Thy all for the great effort!

**Pull request merged**